src/components/external-apis/routes.ts

```diff
@@ -0,0 +1,130 @@
+import { FastifyRequest, FastifyReply } from 'fastify'
+import path from 'path'
+import config from '../../config'
+import { get as extractProperty } from 'lodash'
+import { getPermissionNamesFromJWT } from '../data_display/helpers'
+import axios, { AxiosError, AxiosRequestConfig } from 'axios'
+import { constructAuthHeader, constructQueryObject, validateResult } from './helpers'
+import { ExternalApiConfigs, QueryParameters, PostRoute } from './types'
+import { getApplicationData } from '../actions'
+import { getUserInfo } from '../permissions/loginHelpers'
+import { ActionApplicationData } from '../../types'
+import functions from '../actions/evaluatorFunctions'
+import { errorMessage } from '../utilityFunctions'
+
+interface ExternalApiRequest {
+  name: string
+  route: string
+}
+
+const apiConfigs: ExternalApiConfigs = config?.externalApiConfigs ?? {}
+
+export const routeAccessExternalApi = async (
+  request: FastifyRequest & { auth?: { userId?: number; orgId?: number } },
+  reply: FastifyReply
+) => {
+  const { name, route } = request.params as ExternalApiRequest
+
+  const { baseUrl, routes, authentication } = apiConfigs?.[name]
+  if (!baseUrl) {
+    throw new Error('No baseUrl defined for API: ' + name)
+  }
+
+  const routeConfig = routes?.[route]
+  if (!routeConfig) {
+    throw new Error('No config definition for Route: ' + route)
+  }
+
+  const {
+    method,
+    url,
+    permissions,
+    queryParams,
+    allowedClientQueries,
+    returnProperty,
+    additionalAxiosProperties,
+    validationExpression,
+  } = routeConfig
+
+  if (permissions) {
+    const { permissionNames } = await getPermissionNamesFromJWT(request)
+    const validPermissions = permissions.filter((permission) =>
+      permissionNames.includes(permission)
+    )
+
+    if (validPermissions.length === 0) {
+      reply.status(403)
+      return reply.send('Not authorized to view this resource')
+    }
+  }
+
+  // Construct data object for subsequent expression evaluator
+  const { userId, orgId } = request.auth as { userId?: number; orgId?: number }
```

Comment:
do we need 'as' here, it looks like the type should already have this as per method header

Reply:
It's just that `request.auth` can be undefined, so it then complains about the destructuring. But in reality, it couldn't actually be undefined by this point, as it's checked upstream and only authenticated requests even get this far.

I've changed it to this, which works, albeit adding a tiny bit of runtime overhead vs purely compilational:
```
const { userId, orgId } = request.auth ?? {}
```